tests/lib/vmi.go

```diff
@@ -0,0 +1,332 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package lib
+
+import (
+	"encoding/base64"
+
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	kvirtv1 "kubevirt.io/client-go/api/v1"
+)
+
+type vmiOptions struct {
```

Comment:
Why do you add this level of indirection? We can use `With*` functions to directly manipulate `VirtualMachine`.

Reply:
`New` would just bloat when we add more and more options.

Reply:
> Why do you add this level of indirection? We can use `With*` functions to directly manipulate `VirtualMachine`.

You will have to say more than that for me to understand.
The `With*` pattern here builds a collection of properties with a single interface: `Option` (which is a function with the `*vmiOptions` as its argument). The collection is then processed by the constructor by calling the functions which in turn populates the options.

If you have another proposal, please elaborate (i.e. provide an example).

BTW, in no way this is my invention.

Reply:
> `New` would just bloat when we add more and more options.

Yes, I was looking to have a dedicated namespace for VMI and potentially even extract out the storage parts, but I failed to name the package by the file name. Perhaps you know how to make this happen.
I want to have under `lib/` multiple namespaces for each group.
Then we can do this: `libvmi.New()`.

Reply:
In the current code, `With*` functions are modifying `defaultVMIOptions`. Those options are then used to control modifications on `vmi`.

I suggest you use `With*` callbacks to manipulate `vmi` directly. e.g. the callback returned by `WithCloudInitConfigDriveNetworkData` would take `vmi` as an attribute and set its parameters properly to include the cloud init section.

This way the `New` function does not need to carry the modification logic, only execute callbacks. And writing/reading og those callback `With*` functions would be more intuitive IMO.

The only drawback I see is that this way of handling would not allow you to easily handle conflicts and dependencies. But maybe it would not be an issue if all callbacks are idempotent.

Reply:
You are suggesting a plug-able design. It would have been very nice, especially the notion that every operation is self contained and one can add operations without touching the others (unless dependencies exists).

Unfortunately, this seems to be challenging due to the dependencies between the operation.
- The order is important for some options but the collection is unordered.
- Other options collide with each other.

We could change the options structure to include a handler function, but I am not sure if it is worth it.
I suggest we start with one refactoring and consider the next one per the outcome.

Reply:
@EdDev @phoracek Regarding the file hierarchy, what about "tests/testlibs/libvmi/*.go"?

Reply:
I had a quick discussion on this with @qinqon and we negotiated this: `tests/libvmi/vmi.go`
It seems similar to your suggestion here except the grouping of all helpers.
I think we do not need it now, but if we will, then change will be minimal in the future.

What do you think?

Reply:
@EdDev Sure, also LGTM

Reply:
@EdDev sorry to keep bringing this up, but would you raise one example when my suggestion would not work? For better sleep.